In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Recap of Data Science Problem:
For the following capstone, we aim to build a predicitive model to determine whether a hotel booking will be cancelled. Cancellations are crucial for hotels as it affects revenue and disrupts their operational planning and flow. 

# Overview of Data Set:
The dataset presents a multitude of booking-related features, including lead time, deposit type, and special requests, contributing to the model's intricacy. The primary hurdle resides in the preprocessing phase, necessitating tasks such as feature selection, engineering, addressing missing values, and mitigating data noise. Furthermore, our approach involves training diverse models, assessing their efficacy using appropriate metrics, and make clear model insights through the examination of pivotal features within the context of hotel booking cancellations.

# Variable Descriptions

**Variables and Descriptions**

    hotel:	Type of hotel (Resort Hotel, City Hotel)
    is_canceled: Reservation cancellation status (0 = not canceled, 1 = canceled)
    lead_time: Number of days between booking and arrival
    arrival_date_year: Year of arrival
    arrival_date_month: Month of arrival
    arrival_date_week_number: Week number of the year for arrival
    arrival_date_day_of_month: Day of the month of arrival
    stays_in_weekend_nights: Number of weekend nights (Saturday and Sunday) the guest stayed or booked
    stays_in_week_nights: Number of week nights the guest stayed or booked
    adults: Number of adults
    children: Number of children
    babies: Number of babies
    meal: Type of meal booked (BB, FB, HB, SC, Undefined)
    country: Country of origin of the guest
    market_segment: Market segment designation
    distribution_channel: Booking distribution channel
    is_repeated_guest: If the guest is a repeat customer (0 = not repeated, 1 = repeated)
    previous_cancellations: Number of previous bookings that were canceled by the customer
    previous_bookings_not_canceled: Number of previous bookings that were not canceled by the customer
    reserved_room_type: Type of reserved room
    assigned_room_type: Type of assigned room
    booking_changes: Number of changes made to the booking
    deposit_type: Type of deposit made (No Deposit, Refundable, Non Refund)
    agent: ID of the travel agent responsible for the booking
    company: ID of the company responsible for the booking
    days_in_waiting_list: Number of days the booking was in the waiting list
    customer_type: Type of customer (Transient, Contract, Transient-Party, Group)
    adr: Average Daily Rate
    required_car_parking_spaces: Number of car parking spaces required
    total_of_special_requests: Number of special requests made
    reservation_status: Last reservation status (Check-Out, Canceled, No-Show)
    reservation_status_date: Date of the last reservation status

# Loading the Dataset

In [6]:
hotel_demand = pd.read_csv('hotel_bookings2.csv')
hotel_demand.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  deposit_type  \
0                        0                     0       2  ...    No Deposit   
1                        0                     0       2  ...    No Deposit   
2                        0                     1       1  ...    No Deposit   
3                        0                     1       1  ...    No Deposit   
4                        0                     2       2  ...    No Deposit   

   agent company days_in_waiting_list customer_type   adr  \
0    NaN     NaN                    0     Transient   0.0   
1    NaN     NaN                    0     Transient   0.0   
2    NaN     NaN                    0     Transient  75.0   
3  304.0     NaN                    0     Transient  75.0   
4  240.0     NaN                    0     Transient  98.0   

   required_car_parking_spaces  total_of_special_requests  reservation_status  \
0                            0                          0           Check-Out   
1                            0                          0           Check-Out   
2                            0                          0           Check-Out   
3                            0                          0           Check-Out   
4                            0                          1           Check-Out   

  reservation_status_date  
0              2015-07-01  
1              2015-07-01  
2              2015-07-02  
3              2015-07-02  
4              2015-07-03  

[5 rows x 32 columns]

Lets take a look into the variables we have in the data set. 

In [34]:
hotel_demand.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'company', 'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [35]:
hotel_demand.shape

(119390, 32)

In [36]:
hotel_demand.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 119390 entries, 0 to 119389
Data columns (total 32 columns):
 #   Column                          Non-Null Count   Dtype  
---  ------                          --------------   -----  
 0   hotel                           119390 non-null  object 
 1   is_canceled                     119390 non-null  int64  
 2   lead_time                       119390 non-null  int64  
 3   arrival_date_year               119390 non-null  int64  
 4   arrival_date_month              119390 non-null  object 
 5   arrival_date_week_number        119390 non-null  int64  
 6   arrival_date_day_of_month       119390 non-null  int64  
 7   stays_in_weekend_nights         119390 non-null  int64  
 8   stays_in_week_nights            119390 non-null  int64  
 9   adults                          119390 non-null  int64  
 10  children                        119386 non-null  float64
 11  babies                          119390 non-null  int64  
 12  meal            

We can see that there are 12 object Dtypes, 4 floats and the rest as integers.
We will load the object Dtypes from the data set:

In [37]:
hotel_demand.select_dtypes(object)

hotel arrival_date_month meal country market_segment  \
0       Resort Hotel               July   BB     PRT         Direct   
1       Resort Hotel               July   BB     PRT         Direct   
2       Resort Hotel               July   BB     GBR         Direct   
3       Resort Hotel               July   BB     GBR      Corporate   
4       Resort Hotel               July   BB     GBR      Online TA   
...              ...                ...  ...     ...            ...   
119385    City Hotel             August   BB     BEL  Offline TA/TO   
119386    City Hotel             August   BB     FRA      Online TA   
119387    City Hotel             August   BB     DEU      Online TA   
119388    City Hotel             August   BB     GBR      Online TA   
119389    City Hotel             August   HB     DEU      Online TA   

       distribution_channel reserved_room_type assigned_room_type  \
0                    Direct                  C                  C   
1                    Direct                  C                  C   
2                    Direct                  A                  C   
3                 Corporate                  A                  A   
4                     TA/TO                  A                  A   
...                     ...                ...                ...   
119385                TA/TO                  A                  A   
119386                TA/TO                  E                  E   
119387                TA/TO                  D                  D   
119388                TA/TO                  A                  A   
119389                TA/TO                  A                  A   

       deposit_type customer_type reservation_status reservation_status_date  
0        No Deposit     Transient          Check-Out              2015-07-01  
1        No Deposit     Transient          Check-Out              2015-07-01  
2        No Deposit     Transient          Check-Out              2015-07-02  
3        No Deposit     Transient          Check-Out              2015-07-02  
4        No Deposit     Transient          Check-Out              2015-07-03  
...             ...           ...                ...                     ...  
119385   No Deposit     Transient          Check-Out              2017-09-06  
119386   No Deposit     Transient          Check-Out              2017-09-07  
119387   No Deposit     Transient          Check-Out              2017-09-07  
119388   No Deposit     Transient          Check-Out              2017-09-07  
119389   No Deposit     Transient          Check-Out              2017-09-07  

[119390 rows x 12 columns]

Seeing how many rows for each hotel type:
'City Hotel' and 'Resort Hotel'

In [21]:
hotel_counts = hotel_demand['hotel'].value_counts()
hotel_counts.head()

City Hotel      79330
Resort Hotel    40060
Name: hotel, dtype: int64

Currently 79330 rows of 'City Hotel' and 40060 rows of 'Resort Hotel'

Next, lets see the describe method for the dataset:

In [58]:
hotel_demand.describe()

is_canceled      lead_time  arrival_date_year  \
count  119390.000000  119390.000000      119390.000000   
mean        0.370416     104.011416        2016.156554   
std         0.482918     106.863097           0.707476   
min         0.000000       0.000000        2015.000000   
25%         0.000000      18.000000        2016.000000   
50%         0.000000      69.000000        2016.000000   
75%         1.000000     160.000000        2017.000000   
max         1.000000     737.000000        2017.000000   

       arrival_date_week_number  arrival_date_day_of_month  \
count             119390.000000              119390.000000   
mean                  27.165173                  15.798241   
std                   13.605138                   8.780829   
min                    1.000000                   1.000000   
25%                   16.000000                   8.000000   
50%                   28.000000                  16.000000   
75%                   38.000000                  23.000000   
max                   53.000000                  31.000000   

       stays_in_weekend_nights  stays_in_week_nights         adults  \
count            119390.000000         119390.000000  119390.000000   
mean                  0.927599              2.500302       1.856403   
std                   0.998613              1.908286       0.579261   
min                   0.000000              0.000000       0.000000   
25%                   0.000000              1.000000       2.000000   
50%                   1.000000              2.000000       2.000000   
75%                   2.000000              3.000000       2.000000   
max                  19.000000             50.000000      55.000000   

            children         babies  is_repeated_guest  \
count  119386.000000  119390.000000      119390.000000   
mean        0.103890       0.007949           0.031912   
std         0.398561       0.097436           0.175767   
min         0.000000       0.000000           0.000000   
25%         0.000000       0.000000           0.000000   
50%         0.000000       0.000000           0.000000   
75%         0.000000       0.000000           0.000000   
max        10.000000      10.000000           1.000000   

       previous_cancellations  previous_bookings_not_canceled  \
count           119390.000000                   119390.000000   
mean                 0.087118                        0.137097   
std                  0.844336                        1.497437   
min                  0.000000                        0.000000   
25%                  0.000000                        0.000000   
50%                  0.000000                        0.000000   
75%                  0.000000                        0.000000   
max                 26.000000                       72.000000   

       booking_changes  days_in_waiting_list            adr  \
count    119390.000000         119390.000000  119390.000000   
mean          0.221124              2.321149     101.831122   
std           0.652306             17.594721      50.535790   
min           0.000000              0.000000      -6.380000   
25%           0.000000              0.000000      69.290000   
50%           0.000000              0.000000      94.575000   
75%           0.000000              0.000000     126.000000   
max          21.000000            391.000000    5400.000000   

       required_car_parking_spaces  total_of_special_requests  
count                119390.000000              119390.000000  
mean                      0.062518                   0.571363  
std                       0.245291                   0.792798  
min                       0.000000                   0.000000  
25%                       0.000000                   0.000000  
50%                       0.000000                   0.000000  
75%                       0.000000                   1.000000  
max                       8.000000                   5.000000

# Checking for Null Values

In [31]:
hotel_demand.isnull().any()

hotel                             False
is_canceled                       False
lead_time                         False
arrival_date_year                 False
arrival_date_month                False
arrival_date_week_number          False
arrival_date_day_of_month         False
stays_in_weekend_nights           False
stays_in_week_nights              False
adults                            False
children                           True
babies                            False
meal                              False
country                            True
market_segment                    False
distribution_channel              False
is_repeated_guest                 False
previous_cancellations            False
previous_bookings_not_canceled    False
reserved_room_type                False
assigned_room_type                False
booking_changes                   False
deposit_type                      False
agent                              True
company                            True


We know there are null values for variables 'children', 'country', 'agent' and 'company'
We have to see how many there are to decide whether to drop null values.

In [32]:
hotel_demand.isnull().sum()

hotel                                  0
is_canceled                            0
lead_time                              0
arrival_date_year                      0
arrival_date_month                     0
arrival_date_week_number               0
arrival_date_day_of_month              0
stays_in_weekend_nights                0
stays_in_week_nights                   0
adults                                 0
children                               4
babies                                 0
meal                                   0
country                              488
market_segment                         0
distribution_channel                   0
is_repeated_guest                      0
previous_cancellations                 0
previous_bookings_not_canceled         0
reserved_room_type                     0
assigned_room_type                     0
booking_changes                        0
deposit_type                           0
agent                              16340
company         

Null Value Count

children: 4
country: 488
agent: 16340
company: 112593

In [33]:
missing = pd.concat([hotel_demand.isnull().sum(), 100 * hotel_demand.isnull().mean()], axis=1)
missing.columns=['count', '%']
missing.sort_values(by='%', ascending=False)

count          %
company                         112593  94.306893
agent                            16340  13.686238
country                            488   0.408744
children                             4   0.003350
reserved_room_type                   0   0.000000
assigned_room_type                   0   0.000000
booking_changes                      0   0.000000
deposit_type                         0   0.000000
hotel                                0   0.000000
previous_cancellations               0   0.000000
days_in_waiting_list                 0   0.000000
customer_type                        0   0.000000
adr                                  0   0.000000
required_car_parking_spaces          0   0.000000
total_of_special_requests            0   0.000000
reservation_status                   0   0.000000
previous_bookings_not_canceled       0   0.000000
is_repeated_guest                    0   0.000000
is_canceled                          0   0.000000
distribution_channel                 0   0.000000
market_segment                       0   0.000000
meal                                 0   0.000000
babies                               0   0.000000
adults                               0   0.000000
stays_in_week_nights                 0   0.000000
stays_in_weekend_nights              0   0.000000
arrival_date_day_of_month            0   0.000000
arrival_date_week_number             0   0.000000
arrival_date_month                   0   0.000000
arrival_date_year                    0   0.000000
lead_time                            0   0.000000
reservation_status_date              0   0.000000

In [47]:
hotel_demand.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type', 'agent',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

# Dropping Duplicates and Null Values

There seems to be a significant amount of null values for the company column with roughly 94% of its data is missing. 
We will drop this variable column.

In [ ]:
hotel_demand[hotel_demand.duplicated(keep=False)]

In [ ]:
hotel_demand.drop(columns=['company'], inplace=True)

In [48]:
hotel_demand.drop(columns=['agent'], inplace=True)

In [49]:
hotel_demand.columns

Index(['hotel', 'is_canceled', 'lead_time', 'arrival_date_year',
       'arrival_date_month', 'arrival_date_week_number',
       'arrival_date_day_of_month', 'stays_in_weekend_nights',
       'stays_in_week_nights', 'adults', 'children', 'babies', 'meal',
       'country', 'market_segment', 'distribution_channel',
       'is_repeated_guest', 'previous_cancellations',
       'previous_bookings_not_canceled', 'reserved_room_type',
       'assigned_room_type', 'booking_changes', 'deposit_type',
       'days_in_waiting_list', 'customer_type', 'adr',
       'required_car_parking_spaces', 'total_of_special_requests',
       'reservation_status', 'reservation_status_date'],
      dtype='object')

In [50]:
hotel_demand.shape

(119390, 30)

We made sure that the two variables have been dropped for the upcoming EDA analysis and modeling. 

# Changing the null values from 'country' into 0 values

In [52]:
hotel_demand['country'].fillna(hotel_demand['country'].mode()[0],inplace=True)

In [59]:
hotel_demand.isna().sum()

hotel                             0
is_canceled                       0
lead_time                         0
arrival_date_year                 0
arrival_date_month                0
arrival_date_week_number          0
arrival_date_day_of_month         0
stays_in_weekend_nights           0
stays_in_week_nights              0
adults                            0
children                          4
babies                            0
meal                              0
country                           0
market_segment                    0
distribution_channel              0
is_repeated_guest                 0
previous_cancellations            0
previous_bookings_not_canceled    0
reserved_room_type                0
assigned_room_type                0
booking_changes                   0
deposit_type                      0
days_in_waiting_list              0
customer_type                     0
adr                               0
required_car_parking_spaces       0
total_of_special_requests   

# Checking for other inconsistencies 

In [54]:
hotel_demand['hotel'].value_counts()

City Hotel      79330
Resort Hotel    40060
Name: hotel, dtype: int64

In [57]:
hotel_demand['meal'].value_counts()

BB           92310
HB           14463
SC           10650
Undefined     1169
FB             798
Name: meal, dtype: int64

In [64]:
hotel_demand['meal'].unique()

array(['BB', 'FB', 'HB', 'SC', 'Undefined'], dtype=object)

We can see that there is an undefined value, but we will keep it as is and not change this part of the data.

In [65]:
hotel_demand['children'].value_counts()

0.0     110796
1.0       4861
2.0       3652
3.0         76
10.0         1
Name: children, dtype: int64

In [68]:
hotel_demand['babies'].value_counts()

0     118473
1        900
2         15
10         1
9          1
Name: babies, dtype: int64

In [72]:
hotel_demand.sort_values('babies', ascending=False)

hotel  is_canceled  lead_time  arrival_date_year  \
46619     City Hotel            0         37               2016   
78656     City Hotel            0         11               2015   
94063     City Hotel            0          4               2016   
104351    City Hotel            0         19               2017   
33332   Resort Hotel            0         31               2017   
...              ...          ...        ...                ...   
40138     City Hotel            0          3               2015   
40137     City Hotel            0         71               2015   
40136     City Hotel            0          3               2015   
40135     City Hotel            0         71               2015   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
46619             January                         3   
78656             October                        42   
94063                July                        31   
104351            January                         2   
33332            February                         8   
...                   ...                       ...   
40138                July                        28   
40137                July                        28   
40136                July                        28   
40135                July                        28   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
46619                          12                        0   
78656                          11                        2   
94063                          27                        0   
104351                          8                        2   
33332                          19                        1   
...                           ...                      ...   
40138                          11                        0   
40137                          11                        0   
40136                          11                        0   
40135                          11                        0   
119389                         29                        2   

        stays_in_week_nights  adults  ...  assigned_room_type  \
46619                      2       2  ...                   D   
78656                      1       1  ...                   B   
94063                      4       2  ...                   D   
104351                     2       2  ...                   E   
33332                      0       2  ...                   C   
...                      ...     ...  ...                 ...   
40138                      1       3  ...                   A   
40137                      1       1  ...                   A   
40136                      1       2  ...                   A   
40135                      1       1  ...                   A   
119389                     7       2  ...                   A   

        booking_changes deposit_type days_in_waiting_list    customer_type  \
46619                 1   No Deposit                    0        Transient   
78656                 1   No Deposit                    0  Transient-Party   
94063                 2   No Deposit                    0        Transient   
104351                1   No Deposit                    0        Transient   
33332                 2   No Deposit                    0  Transient-Party   
...                 ...          ...                  ...              ...   
40138                 0   No Deposit                    0  Transient-Party   
40137                 1   No Deposit                    0        Transient   
40136                 0   No Deposit                    0  Transient-Party   
40135                 1   No Deposit                    0        Transient   
119389                0   No Deposit                    0        Transient   

           adr  required_car_parking_spaces  total_of_special_requests  \
46619    84.

In [73]:
hotel_demand.sort_values('children', ascending=False)

hotel  is_canceled  lead_time  arrival_date_year  \
328     Resort Hotel            1         55               2015   
116832    City Hotel            0         14               2017   
40984     City Hotel            0          1               2015   
119070    City Hotel            0          0               2017   
106387    City Hotel            0          3               2017   
...              ...          ...        ...                ...   
119389    City Hotel            0        205               2017   
40600     City Hotel            1          2               2015   
40667     City Hotel            1          1               2015   
40679     City Hotel            1          1               2015   
41160     City Hotel            1          8               2015   

       arrival_date_month  arrival_date_week_number  \
328                  July                        29   
116832               July                        30   
40984              August                        33   
119070             August                        35   
106387           February                         8   
...                   ...                       ...   
119389             August                        35   
40600              August                        32   
40667              August                        32   
40679              August                        32   
41160              August                        33   

        arrival_date_day_of_month  stays_in_weekend_nights  \
328                            12                        4   
116832                         24                        1   
40984                          10                        1   
119070                         29                        0   
106387                         22                        0   
...                           ...                      ...   
119389                         29                        2   
40600                           3                        1   
40667                           5                        0   
40679                           5                        0   
41160                          13                        2   

        stays_in_week_nights  adults  ...  assigned_room_type  \
328                       10       2  ...                   D   
116832                     1       0  ...                   A   
40984                      1       0  ...                   B   
119070                     1       2  ...                   G   
106387                     3       2  ...                   F   
...                      ...     ...  ...                 ...   
119389                     7       2  ...                   A   
40600                      0       2  ...                   B   
40667                      2       2  ...                   B   
40679                      2       3  ...                   B   
41160                      5       2  ...                   B   

        booking_changes deposit_type days_in_waiting_list    customer_type  \
328                   2   No Deposit                    0         Contract   
116832                1   No Deposit                    0  Transient-Party   
40984                 1   No Deposit                    0  Transient-Party   
119070                3   No Deposit                    0            Group   
106387                1   No Deposit                    0        Transient   
...                 ...          ...                  ...              ...   
119389                0   No Deposit                    0        Transient   
40600                 0   No Deposit                    0  Transient-Party   
40667                 0   No Deposit                    0  Transient-Party   
40679                 0   No Deposit                    0  Transient-Party   
41160                 0   No Deposit                    0  Transient-Party   

           adr  required_car_parking_spaces  total_of_special_requests  \
328     133.

In [75]:
filter = (hotel_demand['adults'] == 0) & (hotel_demand['children'] == 0) & (hotel_demand['babies'] == 0)
hotel_demand[filter]

hotel  is_canceled  lead_time  arrival_date_year  \
2224    Resort Hotel            0          1               2015   
2409    Resort Hotel            0          0               2015   
3181    Resort Hotel            0         36               2015   
3684    Resort Hotel            0        165               2015   
3708    Resort Hotel            0        165               2015   
...              ...          ...        ...                ...   
115029    City Hotel            0        107               2017   
115091    City Hotel            0          1               2017   
116251    City Hotel            0         44               2017   
116534    City Hotel            0          2               2017   
117087    City Hotel            0        170               2017   

       arrival_date_month  arrival_date_week_number  \
2224              October                        41   
2409              October                        42   
3181             November                        47   
3684             December                        53   
3708             December                        53   
...                   ...                       ...   
115029               June                        26   
115091               June                        26   
116251               July                        28   
116534               July                        28   
117087               July                        30   

        arrival_date_day_of_month  stays_in_weekend_nights  \
2224                            6                        0   
2409                           12                        0   
3181                           20                        1   
3684                           30                        1   
3708                           30                        2   
...                           ...                      ...   
115029                         27                        0   
115091                         30                        0   
116251                         15                        1   
116534                         15                        2   
117087                         27                        0   

        stays_in_week_nights  adults  ...  assigned_room_type  \
2224                       3       0  ...                   I   
2409                       0       0  ...                   I   
3181                       2       0  ...                   C   
3684                       4       0  ...                   A   
3708                       4       0  ...                   C   
...                      ...     ...  ...                 ...   
115029                     3       0  ...                   A   
115091                     1       0  ...                   K   
116251                     1       0  ...                   K   
116534                     5       0  ...                   K   
117087                     2       0  ...                   A   

        booking_changes deposit_type days_in_waiting_list    customer_type  \
2224                  1   No Deposit                    0  Transient-Party   
2409                  0   No Deposit                    0        Transient   
3181                  0   No Deposit                    0  Transient-Party   
3684                  1   No Deposit                  122  Transient-Party   
3708                  1   No Deposit                  122  Transient-Party   
...                 ...          ...                  ...              ...   
115029                1   No Deposit                    0        Transient   
115091                0   No Deposit                    0        Transient   
116251                2   No Deposit                    0        Transient   
116534                1   No Deposit                    0  Transient-Party   
117087                0   No Deposit                    0        Transient   

           adr  required_car_parking_spaces  total_of_special_requests  \
2224      0.

From the above code, we can see that there are 180 rows of data where all adults, children and babies are 0. This does not make sense, so we will drop these rows. 

In [70]:
hotel_demand_cleaned = hotel_demand[~filter]
hotel_demand_cleaned

hotel  is_canceled  lead_time  arrival_date_year  \
0       Resort Hotel            0        342               2015   
1       Resort Hotel            0        737               2015   
2       Resort Hotel            0          7               2015   
3       Resort Hotel            0         13               2015   
4       Resort Hotel            0         14               2015   
...              ...          ...        ...                ...   
119385    City Hotel            0         23               2017   
119386    City Hotel            0        102               2017   
119387    City Hotel            0         34               2017   
119388    City Hotel            0        109               2017   
119389    City Hotel            0        205               2017   

       arrival_date_month  arrival_date_week_number  \
0                    July                        27   
1                    July                        27   
2                    July                        27   
3                    July                        27   
4                    July                        27   
...                   ...                       ...   
119385             August                        35   
119386             August                        35   
119387             August                        35   
119388             August                        35   
119389             August                        35   

        arrival_date_day_of_month  stays_in_weekend_nights  \
0                               1                        0   
1                               1                        0   
2                               1                        0   
3                               1                        0   
4                               1                        0   
...                           ...                      ...   
119385                         30                        2   
119386                         31                        2   
119387                         31                        2   
119388                         31                        2   
119389                         29                        2   

        stays_in_week_nights  adults  ...  assigned_room_type  \
0                          0       2  ...                   C   
1                          0       2  ...                   C   
2                          1       1  ...                   C   
3                          1       1  ...                   A   
4                          2       2  ...                   A   
...                      ...     ...  ...                 ...   
119385                     5       2  ...                   A   
119386                     5       3  ...                   E   
119387                     5       2  ...                   D   
119388                     5       2  ...                   A   
119389                     7       2  ...                   A   

        booking_changes deposit_type days_in_waiting_list customer_type  \
0                     3   No Deposit                    0     Transient   
1                     4   No Deposit                    0     Transient   
2                     0   No Deposit                    0     Transient   
3                     0   No Deposit                    0     Transient   
4                     0   No Deposit                    0     Transient   
...                 ...          ...                  ...           ...   
119385                0   No Deposit                    0     Transient   
119386                0   No Deposit                    0     Transient   
119387                0   No Deposit                    0     Transient   
119388                0   No Deposit                    0     Transient   
119389                0   No Deposit                    0     Transient   

           adr  required_car_parking_spaces  total_of_special_requests  \
0         0.00                            0     

Lets save it as df.

In [76]:
df = hotel_demand_cleaned

In [77]:
df.head()

hotel  is_canceled  lead_time  arrival_date_year arrival_date_month  \
0  Resort Hotel            0        342               2015               July   
1  Resort Hotel            0        737               2015               July   
2  Resort Hotel            0          7               2015               July   
3  Resort Hotel            0         13               2015               July   
4  Resort Hotel            0         14               2015               July   

   arrival_date_week_number  arrival_date_day_of_month  \
0                        27                          1   
1                        27                          1   
2                        27                          1   
3                        27                          1   
4                        27                          1   

   stays_in_weekend_nights  stays_in_week_nights  adults  ...  \
0                        0                     0       2  ...   
1                        0                     0       2  ...   
2                        0                     1       1  ...   
3                        0                     1       1  ...   
4                        0                     2       2  ...   

   assigned_room_type  booking_changes deposit_type days_in_waiting_list  \
0                   C                3   No Deposit                    0   
1                   C                4   No Deposit                    0   
2                   C                0   No Deposit                    0   
3                   A                0   No Deposit                    0   
4                   A                0   No Deposit                    0   

  customer_type   adr  required_car_parking_spaces  total_of_special_requests  \
0     Transient   0.0                            0                          0   
1     Transient   0.0                            0                          0   
2     Transient  75.0                            0                          0   
3     Transient  75.0                            0                          0   
4     Transient  98.0                            0                          1   

   reservation_status reservation_status_date  
0           Check-Out              2015-07-01  
1           Check-Out              2015-07-01  
2           Check-Out              2015-07-02  
3           Check-Out              2015-07-02  
4           Check-Out              2015-07-03  

[5 rows x 30 columns]

# Saving file as csv

In [79]:
df.to_csv('df', index=False)